Task Management System

SGrechnev opened the pull request:
Read project-TaskManagementSystem/README.md

SGrechnev committed:
Task Management System
Read project-TaskManagementSystem/README.md

SGrechnev committed:
Fix bugs

- TaskRestController: fix logger creation
- Task: initialize reports for correct JSON translation
- app.js: fix addTask function
- tasks.html and users.html: replace "th:utext" with "th:text" to avoid XSS-injection

petrelevich commented on project-TaskManagementSystem/src/main/resources/schema-postgres.sql:
```diff
@@ -0,0 +1,31 @@
+drop sequence if exists hibernate_sequence;
```
схему стоит накатывать через миграции.

petrelevich commented on project-TaskManagementSystem/src/main/java/ru/otus/service/AuthenticatedUserInfoService.java:
```diff
@@ -0,0 +1,21 @@
+        var ldapUserDetails = (LdapUserDetailsImpl) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
+        String username = ldapUserDetails.getUsername();
+        var optionalUser = userService.findByUsername(username);
+        return optionalUser.orElse(null);
```
на случай null действительно не надо предусмотреть какую-то более тонкую реакцию?

SGrechnev replied:
Переделал. В таком случае бросается исключение и аутентификация считается неуспешной.

petrelevich commented on project-TaskManagementSystem/src/main/java/ru/otus/service/TaskServiceImpl.java:
```diff
@@ -0,0 +1,100 @@
+    private TaskRepository taskRepository;
+
+    @Autowired
+    private UserService userService;
```
инжектировать зависимости лучше через конструктор.

SGrechnev replied:
сделано

petrelevich commented on project-TaskManagementSystem/src/main/java/ru/otus/service/TaskServiceImpl.java:
```diff
@@ -0,0 +1,100 @@
+        }
+        var task = oTask.get();
+        task.setProgress(progress);
+        if(progress == 100) {
```
тут и в похожих местах это 100% ?
лучше использовать констранту.

SGrechnev replied:
сделано

petrelevich commented on project-TaskManagementSystem/src/main/java/ru/otus/service/ReportServiceImpl.java:
```diff
@@ -0,0 +1,48 @@
+            throw new IllegalArgumentException(String.format("Adding report for non-existing task with id %d", reportDto.getTaskId()));
+        }
+        return reportRepository.save(
+                new Report(task.get(), reportDto.getProgress(), reportDto.getComment(), LocalDate.now())
```
это преобразование можно через map добавить к optional task

SGrechnev committed:
Refactor

- dependency injection via constructor
- constants for min and max progress

SGrechnev committed:
Minor changes

- add a common path with different redirects for different roles
- add exception handler and logger

SGrechnev committed:
Front changes

- beautify add-task form and report cards
- add logo to header

SGrechnev committed:
Improve authorization

- add CustomUserDetailsContextMapper to enrich LDAPUser with Role
  from local DB
- add authorization on some paths

SGrechnev committed:
Improvement error handling. Refactoring.

- fix error 'LDAP user not found in local DB' handling
- fix error handling in WebController: return viewname, not message

SGrechnev committed:
Flyway
Initializing the database with flyway

SGrechnev committed:
Update WebSecurityConfig.java

remove commented code

SGrechnev committed:
Refactoring

- delete unused dependencies from build.gradle
- ReportService depends on TaskService, not TaskRepository

SGrechnev committed:
minor changes

- add report via js
- refactor app.js (selectors)
- update bootstrap (4.4.1-1 => 4.6.0)

SGrechnev committed:
fragments

SGrechnev committed:
add notifications in toasts

SGrechnev committed:
LDAP and DB changes

SGrechnev committed:
Minor changes

- handle exceptions on user saving
- refactor: change Test.toString method, russian excepetion messages,
  remove redundant logging

SGrechnev committed:
Add method 'delete task'
